README.md
[![pub package](https://img.shields.io/pub/v/flutter_pannable_rating_bar?color=green&include_prereleases&style=plastic)](https://pub.dev/packages/flutter_pannable_rating_bar)

# PannableRatingBar

Introducing a new and improved rating bar for Flutter, offering versatility and full customization. 
Unlike other rating bars, this one allows for any value to be selected, not just full or half 
ratings when interacted. Tap, drag and hover gestures are seamlessly supported, providing a smooth 
user experience.

## Features
* Offers fractional values for rating
* High degree of customization for the rating widgets, which can be of different sizes, shapes, and colors.
* Precise hit testing, as the child widget's render object is utilized to determine the results.
* Allows for filtering gestures with either `tapOnly`, `dragOnly` or `tapAndDrag` options.
* Rating reports on cursor hovering supported.
* Built on Flutter's `Wrap` widget, offering a variety of supported layouts, including consideration 
of `Wrap` properties such as `textDirection` and `verticalDirection` when painting the indicators.

## Getting started

First import the widget

```dart
import 'package:flutter_pannable_rating_bar/flutter_pannable_rating_bar.dart';
```

## Usage

This widget is stateless and highly flexible, allowing you to easily manipulate values and 
customize the display of the `PannableRatingBar`. Simply rebuild the widget with the value provided 
in the `onChanged` callback or `onHover` callback and the widget will automatically adjust the 
rating distribution across each rating widget.

<img src="https://github.com/SilentCatD/flutter_pannable_rating_bar/blob/main/assets/basics.gif?raw=true" width="200px">

```dart
double rating = 0.0;

PannableRatingBar(
  rate: rating,
  items: List.generate(5, (index) =>
    const RatingWidget(
      selectedColor: Colors.yellow,
      unSelectedColor: Colors.grey,
      child: Icon(
        Icons.star,
        size: 48,
      ),
    )),
  onChanged: (value) { // the rating value is updated on tap or drag.
    setState(() {
      rating = value;
    });
  },
)
```

The widget also supports getting the rate value by hovering with the `onHover` callback. It works 
best on devices that support a mouse cursor, including desktop web browsers.

```dart
double rating = 0.0;

PannableRatingBar(
  rate: rating,
  items: List.generate(5, (index) =>
    const RatingWidget(
      selectedColor: Colors.yellow,
      unSelectedColor: Colors.grey,
      child: Icon(
        Icons.star,
        size: 48,
      ),
    )),
  onHover: (value) { // the rating value is updated every time the cursor moves over a new item.
    setState(() {
      rating = value;
    });
  },
)
```


Utilize your own custom widgets as the rating indicators. With complete customization freedom, they 
don't have to be uniform in size, color, or any other aspect.

<img src="https://github.com/SilentCatD/flutter_pannable_rating_bar/blob/main/assets/crazy.gif?raw=true" width="200px">

```dart
double rating = 0.0;

PannableRatingBar(
  rate: rating,
  onChanged: (value) {
    setState(() {
      rating = value;
    });
  },
  items: const [
    RatingWidget(
      selectedColor: Colors.yellow,
      unSelectedColor: Colors.grey,
      child: Icon(
        Icons.star,
        size: 48,
      ),
    ),
    RatingWidget(
      selectedColor: Colors.blue,
      unSelectedColor: Colors.red,
      child: Icon(
        Icons.ac_unit,
        size: 48,
      ),
    ),
    RatingWidget(
      selectedColor: Colors.purple,
      unSelectedColor: Colors.amber,
      child: Icon(
        Icons.access_time_filled,
        size: 48,
      ),
    ),
    RatingWidget(
      selectedColor: Colors.cyanAccent,
      unSelectedColor: Colors.grey,
      child: Icon(
        Icons.abc,
        size: 48,
      ),
    ),
    RatingWidget(
      selectedColor: Colors.tealAccent,
      unSelectedColor: Colors.purple,
      child: Icon(
        Icons.accessibility_new_sharp,
        size: 48,
      ),
    ),
  ],
)
```

<img src="https://github.com/SilentCatD/flutter_pannable_rating_bar/blob/main/assets/variety.gif?raw=true" width="200px">

```dart
double rating = 0;
PannableRatingBar(
  rate: rating,
  onChanged: (value){
    setState((){
      rating = value;
    });
  },
  spacing: 20,
  items: const [
    RatingWidget(
      selectedColor: Colors.blue,
      child: Text(
        "Pannable",
        style: TextStyle(color: Colors.grey, fontSize: 40),
      ),
    ),
    RatingWidget(
      selectedColor: Colors.red,
      child: Text(
        "Rating",
        style: TextStyle(color: Colors.grey, fontSize: 30),
      ),
    ),
    RatingWidget(
      selectedColor: Colors.amber,
      child: Text(
        "Bar",
        style: TextStyle(color: Colors.grey, fontSize: 50),
      ),
    ),
  ],
),
```


This widget utilizes the layout capabilities of Flutter's built-in `Wrap` widget, allowing for a wide 
range of possible layouts to be achieved. The properties of the `Wrap` widget, such as `textDirection` 
and `verticalDirection`, are also considered when painting the indicators, giving you even greater 
control over the final look and feel of the rating bar.

<img src="https://github.com/SilentCatD/flutter_pannable_rating_bar/blob/main/assets/builder.gif?raw=true" width="200px">

```dart
double rating = 0.0;

PannableRatingBar.builder(
  rate: rating,
  alignment: WrapAlignment.center,
  spacing: 20,
  runSpacing: 10,
  itemCount: 20,
  direction: Axis.vertical,
  itemBuilder: (context, index) {
    return const RatingWidget(
      selectedColor: Colors.yellow,
      unSelectedColor: Colors.grey,
      child: Icon(
        Icons.star,
        size: 48,
      ),
    );
  },
  onChanged: (value) {
    setState(() {
      rating = value;
    });
  },
)
```

Customize the appearance of the rating indicators based on their values.

<img src="https://github.com/SilentCatD/flutter_pannable_rating_bar/blob/main/assets/animation.gif?raw=true" width="200px">

```dart
double rating = 0;

PannableRatingBar(
    rate: rating,
    spacing: 20,
    onChanged: (value){
      setState((){
        rating = value;
      });
    },
    items: [
      RatingWidget(
        selectedColor: rating <= 0.5 ? Colors.red : Colors.green,
        unSelectedColor: Colors.grey,
        child: AnimatedSwitcher(
          duration: const Duration(milliseconds: 300),
          child: rating <= 0.5
              ? const Icon(
                  Icons.sentiment_very_dissatisfied,
                  key: ValueKey("sad"),
                  size: 100,
                )
              : const Icon(
                  Icons.sentiment_satisfied,
                  key: ValueKey("happy"),
                  size: 100,
                ),
        ),
      ),
    ],
),
```

For ease of transforming / mutating rating values, the library support using functions to do as such,
used as a parameter in `PannableRatingBar`.

Return null from this will cause the current rating value skipped in all callbacks.
```dart
typedef RatingValueTransformer = double? Function(double value);
```
Some of the pre-built `RatingValueTransformer`s available are:

```dart
// Rounding to nearest single fractional digit. This is also the default one that 
// `PannableRatingBar` used.
double singleFractionalValueTransformer(double value);
```

```dart
// Rounding to nearest .5.
double halfFractionalValueTransformer(double value);
```

```dart
// Leave the rating as is.
double rawValueTransformer(double value);
```

## Additional information

For more, checkout and play with the example file.
If you have any suggestions for improvement, simply create an issue in the repository!



CHANGELOG.md
# Release Notes

### 2.7.2+1

* Files: update `.pubignore` and assets link.

### 2.7.2

* Update: change pre-built `RatingValueTransformer` function names.

### 2.7.1

* Update: changing impl of internal `ShaderMask` drawing to single color.

### 2.7.0

* New feature: `GestureType.DragOnly` added, allow further gesture filtering.

### 2.6.0

* New feature: `RatingValueTransformer` added, allow transforming rating values.

## 2.5.2+1

* Files: remove `assets` file in publishing.

## 2.5.2

* Update: Add `assert(debugHandleEvent(...))` in the `handleEvent` function according to docs

## 2.5.1+2

* Files: remove `assets` file in publishing.

## 2.5.1+1

* Optimization: reduce library size by ignore `assets` folder.

## 2.5.1

* Update: Remove unused variable `visibleForTracking`.
* Update: Mark `calcPercent` function `protected`.
* Documentation: Added more examples and formatted the `example` file.

## 2.5.0+1

* Documentation: Change package description to fit with pub.dev standard.

## 2.5.0

* New feature: Added `onCompleted` callback, which will reports the rating value at the position
  which user stop contacting the device.

## 2.4.1+1

* Documentation: Fix documentation typo in README.

## 2.4.1

* Update: changing internal implementation of calculate indexes of children.

## 2.4.0

* New feature: Added `onHover` callback, which will reports the rating value of the current cursor
  position.
* Documentation: Adding example for `onHover` use case.

## 2.3.0+2

* Documentation: Changes in package description.

## 2.3.0+1

* Documentation: Small change in README.

## 2.3.0

* New feature: Added ability to filter out gestures - choose between `tapOnly` or `tapAndDrag`.
* New feature: Getter for `itemCount` added. Subclasses of `PannableRatingBar` can now access it.
* Documentation: Most of the docs have been rewritten with GPT under my supervision.

## 2.2.0

* New feature: Added `minRating` and `maxRating` parameters.

## 2.1.2

* Update: `BlendMode` changed to `srcIn` instead of `srcATop`.

## 2.1.1

* Optimization: Skip rate calculation process when `onChanged` callback is null.

## 2.1.0+3

* Documentation: Updated example file.

## 2.1.0+2

* Documentation: Added new gif and more examples.

## 2.1.0+1

* Documentation: Added new gif file and example file.

## 2.1.0

* Documentation: README.md code snippets updated for better comprehension of widget usage.
* Update: Constraint for rate value removed. It can now be set as negative or greater
  than `itemCount`.
* New feature: Support for `Wrap` properties added - `textDirection` and `verticalDirection`. The
  widget now takes these properties into consideration and paints indicators correctly.
* Update: Non-nullable constraint of `RatingWidget.unSelectedColor` removed. It is now nullable, and
  if set to null, background painting will be skipped. The implementation of drawing changed
  from `ShaderMask` to `ColorFilter`.

## 2.0.1

* New feature: Parameter for pixels compensation added. Refer to this
  issue: https://github.com/flutter/flutter/issues/98464.

## 2.0.0

* BREAKING CHANGES!
* New feature: Added support for customizing each rating widget.
* New feature: Added builder constructor support.

## 1.0.0

* Initial release.

## 0.0.1

* Preparation for release.

test/flutter_pannable_rating_bar_test.dart
// import 'package:flutter_test/flutter_test.dart';
//
// import 'package:flutter_pannable_rating_bar/flutter_pannable_rating_bar.dart';
//
// void main() {
// test('adds one to input values', () {
//   final calculator = Calculator();
//   expect(calculator.addOne(2), 3);
//   expect(calculator.addOne(-7), -6);
//   expect(calculator.addOne(0), 1);
// });
// }


example/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_pannable_rating_bar/flutter_pannable_rating_bar.dart';

/// For the [onHover] example, view on Web or desktop related device for best
/// experiences.

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return DefaultTabController(
      length: 6,
      child: Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
          bottom: const TabBar(
            isScrollable: true,
            tabs: [
              Tab(
                text: "Basics",
              ),
              Tab(
                text: "Variety",
              ),
              Tab(
                text: "Customized",
              ),
              Tab(
                text: "Multiple",
              ),
              Tab(
                text: "Animation",
              ),
              Tab(
                text: "Hover",
              )
            ],
          ),
        ),
        body: const TabBarView(
          physics: NeverScrollableScrollPhysics(),
          children: [
            BasicExample(),
            VarietyExample(),
            CustomizedExample(),
            MultipleExample(),
            AnimatedExample(),
            HoverExample(),
          ],
        ),
      ),
    );
  }
}

class BasicExample extends StatefulWidget {
  const BasicExample({Key? key}) : super(key: key);

  @override
  State<BasicExample> createState() => _BasicExampleState();
}

class _BasicExampleState extends State<BasicExample> {
  double rating = 0;

  void updateRating(double value) {
    setState(() {
      rating = value;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Text(
            '$rating',
            style: Theme.of(context).textTheme.headline4,
          ),
          const SizedBox(
            height: 10,
          ),
          PannableRatingBar(
            rate: rating,
            onChanged: updateRating,
            spacing: 20,
            items: List.generate(
              5,
              (index) => const RatingWidget(
                selectedColor: Colors.yellow,
                unSelectedColor: Colors.grey,
                child: Icon(
                  Icons.star,
                  size: 48,
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

class VarietyExample extends StatefulWidget {
  const VarietyExample({Key? key}) : super(key: key);

  @override
  State<VarietyExample> createState() => _VarietyExampleState();
}

class _VarietyExampleState extends State<VarietyExample> {
  double rating = 0;

  void updateRating(double value) {
    setState(() {
      rating = value;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(16),
      child: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              '$rating',
              style: Theme.of(context).textTheme.headline4,
            ),
            const SizedBox(
              height: 10,
            ),
            PannableRatingBar(
              rate: rating,
              onChanged: updateRating,
              spacing: 20,
              items: const [
                RatingWidget(
                  selectedColor: Colors.blue,
                  child: Text(
                    "Pannable",
                    style: TextStyle(color: Colors.grey, fontSize: 40),
                  ),
                ),
                RatingWidget(
                  selectedColor: Colors.red,
                  child: Text(
                    "Rating",
                    style: TextStyle(color: Colors.grey, fontSize: 30),
                  ),
                ),
                RatingWidget(
                  selectedColor: Colors.amber,
                  child: Text(
                    "Bar",
                    style: TextStyle(color: Colors.grey, fontSize: 50),
                  ),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

class CustomizedExample extends StatefulWidget {
  const CustomizedExample({Key? key}) : super(key: key);

  @override
  State<CustomizedExample> createState() => _CustomizedExampleState();
}

class _CustomizedExampleState extends State<CustomizedExample> {
  double rating = 0;

  void updateRating(double value) {
    setState(() {
      rating = value;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Text(
            '$rating',
            style: Theme.of(context).textTheme.headline4,
          ),
          const SizedBox(
            height: 10,
          ),
          PannableRatingBar(
            rate: rating,
            spacing: 20,
            onChanged: updateRating,
            items: const [
              RatingWidget(
                selectedColor: Colors.yellow,
                unSelectedColor: Colors.grey,
                child: Icon(
                  Icons.star,
                  size: 48,
                ),
              ),
              RatingWidget(
                selectedColor: Colors.blue,
                unSelectedColor: Colors.red,
                child: Icon(
                  Icons.ac_unit,
                  size: 48,
                ),
              ),
              RatingWidget(
                selectedColor: Colors.purple,
                unSelectedColor: Colors.amber,
                child: Icon(
                  Icons.access_time_filled,
                  size: 48,
                ),
              ),
              RatingWidget(
                selectedColor: Colors.cyanAccent,
                unSelectedColor: Colors.grey,
                child: Icon(
                  Icons.abc,
                  size: 48,
                ),
              ),
              RatingWidget(
                selectedColor: Colors.tealAccent,
                unSelectedColor: Colors.purple,
                child: Icon(
                  Icons.accessibility_new_sharp,
                  size: 48,
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }
}

class MultipleExample extends StatefulWidget {
  const MultipleExample({Key? key}) : super(key: key);

  @override
  State<MultipleExample> createState() => _MultipleExampleState();
}

class _MultipleExampleState extends State<MultipleExample> {
  double rating = 0;

  void updateRating(double value) {
    setState(() {
      rating = value;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Expanded(
            child: Center(
              child: Text(
                '$rating',
                style: Theme.of(context).textTheme.headline4,
                textAlign: TextAlign.center,
              ),
            ),
          ),
          Expanded(
            child: PannableRatingBar.builder(
              rate: rating,
              alignment: WrapAlignment.center,
              spacing: 20,
              runSpacing: 10,
              itemCount: 20,
              direction: Axis.vertical,
              textDirection: TextDirection.rtl,
              verticalDirection: VerticalDirection.up,
              onChanged: updateRating,
              itemBuilder: (context, index) {
                return const RatingWidget(
                  selectedColor: Colors.yellow,
                  unSelectedColor: Colors.grey,
                  child: Icon(
                    Icons.star,
                    size: 48,
                  ),
                );
              },
            ),
          )
        ],
      ),
    );
  }
}

class AnimatedExample extends StatefulWidget {
  const AnimatedExample({Key? key}) : super(key: key);

  @override
  State<AnimatedExample> createState() => _AnimatedExampleState();
}

class _AnimatedExampleState extends State<AnimatedExample> {
  double rating = 0;

  void updateRating(double value) {
    setState(() {
      rating = double.parse(value.toStringAsFixed(1));
    });
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          PannableRatingBar(
            rate: rating,
            spacing: 20,
            onChanged: updateRating,
            items: [
              RatingWidget(
                selectedColor: rating <= 0.5 ? Colors.red : Colors.green,
                unSelectedColor: Colors.grey,
                child: AnimatedSwitcher(
                  duration: const Duration(milliseconds: 300),
                  child: rating <= 0.5
                      ? const Icon(
                          Icons.sentiment_very_dissatisfied,
                          key: ValueKey("sad"),
                          size: 100,
                        )
                      : const Icon(
                          Icons.sentiment_satisfied,
                          key: ValueKey("happy"),
                          size: 100,
                        ),
                ),
              ),
            ],
          ),
          const SizedBox(
            height: 10,
          ),
          Slider(
            value: rating,
            onChanged: updateRating,
            min: 0,
            max: 1,
          ),
        ],
      ),
    );
  }
}

class HoverExample extends StatefulWidget {
  const HoverExample({Key? key}) : super(key: key);

  @override
  State<HoverExample> createState() => _HoverExampleState();
}

class _HoverExampleState extends State<HoverExample> {
  double rating = 0;
  double ratingSet = 0;

  void updateRating(double value) {
    setState(() {
      rating = value;
    });
  }

  void setRating(double value) {
    setState(() {
      ratingSet = value;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Text(
            '$ratingSet',
            style: Theme.of(context).textTheme.headline4,
          ),
          const SizedBox(
            height: 10,
          ),
          PannableRatingBar(
            rate: rating,
            onChanged: setRating,
            onHover: updateRating,
            spacing: 20,
            items: List.generate(
              5,
              (index) => const RatingWidget(
                selectedColor: Colors.yellow,
                unSelectedColor: Colors.grey,
                child: Icon(
                  Icons.star,
                  size: 48,
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
